pipeline/psth.py

```diff
@@ -19,129 +22,229 @@
+                    'outcome': 'miss',
+                    'early_lick': 'no early',
+                    'trial_instruction': 'left'}
+            },
+            {
+                'trial_condition_desc': 'good_noearlylick_right_miss',
+                'trial_condition_func': '_get_trials_no_stim',
+                'trial_condition_arg': {
+                    'task': 'audio delay',
+                    'task_protocol': 1,
+                    'outcome': 'miss',
+                    'early_lick': 'no early',
+                    'trial_instruction': 'right'}
+            },
+        )
+
+        return ({**d, 'trial_condition_id': key_hash(d['trial_condition_arg'])}
+                for d in contents_data)
+
+    @classmethod
+    def get_trials(cls, trial_condition_desc):
+        return cls.get_func({'trial_condition_desc': trial_condition_desc})()
+
+    @classmethod
+    def get_func(cls, key):
+        self = cls()
+
+        func, args = (self & key).fetch1(
+            'trial_condition_func', 'trial_condition_arg')
+
+        return partial(dict(getmembers(cls))[func], **args)
+
+    @classmethod
+    def _get_trials_no_stim(cls, task=None, task_protocol=None, outcome=None,
+                            early_lick=None, trial_instruction=None):
+
+        log.debug('_get_trials_no_stim', locals())
+
+        return ((experiment.BehaviorTrial
+                 & {'task': task}
+                 & {'trial_instruction': trial_instruction}
+                 & {'early_lick': early_lick}
+                 & {'outcome': outcome}) - experiment.PhotostimEvent)
+
+    @classmethod
+    def _get_trials_stim(cls, task=None, task_protocol=None, outcome=None,
+                         early_lick=None, trial_instruction=None):
+
+        log.debug('_get_trials_stim', locals())
+
+        return ((experiment.BehaviorTrial
+                 & {'task': task}
+                 & {'trial_instruction': trial_instruction}
+                 & {'early_lick': early_lick}
+                 & {'outcome': outcome}) & experiment.PhotostimEvent)
+
+
+@schema
+class UnitPsth(dj.Computed):
+    definition = """
+    -> TrialCondition
+    -> ephys.Unit
+    ---
+    unit_psth=NULL:                             longblob
+    """
     psth_params = {'xmin': -3, 'xmax': 3, 'binsize': 0.04}
 
-    @staticmethod
-    def compute_unit_trial_psth(unit_key, trial_keys):
+    key_source = (ephys.Unit * TrialCondition
+                  & [{'trial_condition_desc': desc}
+                     for desc in ['good_noearlylick_left_hit',
+                                  'good_noearlylick_right_hit',
+                                  'good_noearlylick_left_miss',
+                                  'good_noearlylick_right_miss']])
 
-        q = (ephys.TrialSpikes() & unit_key & trial_keys)
+    def make(self, key):
+        log.info('UnitPsth.make(): key: {}'.format(key))
+
+        # build unit key,
+        unit = {k: v for k, v in key.items() if k in ephys.Unit.primary_key}
```

Comment:
There should be no need for this to be restricted down to the unit key - just use `key` as is in the restriction.

Reply:
fixed